### ISSUE TITLE
Crackling noise from annoying to inaudible

### ISSUE THREAD

**Issue by LeoBoudet** [LMMS/lmms#5418]
### Bug Summary
I know this issue seems quite known in Linux world but I can't find exactly the same description as my own problem. When I open LMMS, I can't play any sound without having a monstruous crackling noise. After a few minutes it gets better but it keeps producing a clicking sound at the beginning of each sampled note (first I thought it was clipping but modify the volume or the form of the wave or whatever doesn't change anything). It is more obvious on some instruments than others but it seems to concern everyone of them.

I tried with different speakers and different softwares, the problem concerns uniquely LMMS and my audio set works perfectly as far as I know.

The sound configuration in LMMS is set on ALSA (default).

#### Steps to reproduce
Open LMMS and play any note, wait a bit and keep playing to hear if the bug evolves.

#### Expected behavior
Normal sound without crackling and clicking noises.

#### Actual behavior
Crackling and clicking noises.

#### Affected LMMS versions
Version 1.2.1 (Linux/x86_64, Qt 5.9.7, GCC 4.8.4) (appimage) on Ubuntu 18.04.4 LTS, Gnome environment.

Thanks for your help!

Léo

**Comment by musikBear** [comment]
> The sound configuration in LMMS is set on ALSA (default).

Could you retry with SDL?
SDL work similarly fine in both win and linux, so a test with SDL will point the issue to you ALSA backend, or clear it.

**Comment by LeoBoudet** [comment]
The awful crackling noise didn't come up again since a moment so I can't tell the difference on that particular point but for the clicking noise, it's exactly the same results with SDL or with ALSA.

**Comment by LeoBoudet** [comment]
After many tests I noticed that the clicking noise appears sometimes just at the beginning of the note, sometimes at the end, sometimes both, sometimes not. It depends on the instrument apparently. Maybe it's a sort of clipping that I couldn't solve until now?

I tried to play some demos and there is no clicking sound.

There is quite known horrible crackling sound issue with the based-debian distros. Maybe my problem is a mix of that and a matter of settings in LMMS?

**Comment by musikBear** [comment]
> The awful crackling noise didn't come up again since a moment

How long ago is it since you made a cold-boot of your system?
Windows at least, can 'act up' and make artefacts, if it is not restarted 'now and then'. I believe linux is less prone to that issue, but you should try any way.
Clicking noise is mostly because of missing envelopes. 

Retry with a real preset.
I recommend anything from zynAddSubFX, they are rock-solid presets! **001-saw strings**
What is the name of the plugin that causes these click/pops ?
You have to understand, that this issue is not general. It is yours only, so we are more in 'help-sesk' territory, that bug resolve

**Comment by LeoBoudet** [comment]
> You have to understand, that this issue is not general. It is yours only, so we are more in 'help-sesk' territory, that bug resolve

Ok, should I go to the forum and close the issue here?

**Comment by SecondFlight** [comment]
I think this might be the same issue as #3086.

@LeoBoudet Could you record the output of LMMS with Audacity when this bug occurs, and say which instrument(s) you're using? And if you haven't already, does setting an envelope with a small amount of attack and release fix the issue?

**Comment by LeoBoudet** [comment]
> Clicking noise is mostly because of missing envelopes.

>Retry with a real preset.
>I recommend anything from zynAddSubFX, they are rock-solid presets! 001-saw strings
>What is the name of the plugin that causes these click/pops ?

I think you're probably right. It is better with zynAddSubFX.
The plugins that cause the clicking sounds are: the default TripleOscillator and kicker of the bassline, Organic, Opulenz, sfxr, vibed... and probably some others.

> I think this might be the same issue as #3086.
Ah, I didn't find this issue before posting mine, my bad. Looks pretty similar indeed.

Hear in the file attached: TripleOscillator and kicker. The kicker is set with a tiny attack and release. One of the sound of TripleOscillator is the one by default (second), the first is set on very low attack & release for the three oscillators.

[clicking-noises.mp3.zip](https://github.com/LMMS/lmms/files/4330607/clicking-noises.mp3.zip)

**Comment by SecondFlight** [comment]
That sounds like the typical clicking you get with the envelope disabled. Did you click to enable the envelope before setting it? The envelope should be green instead of grey.

**Comment by SecondFlight** [comment]
![image](https://user-images.githubusercontent.com/6700184/76643399-527acc80-652b-11ea-9f1c-0552052cb242.png)

**Comment by musikBear** [comment]
https://i.imgur.com/sdJcMaf.png
This what secondflight means with enabled envelope and green instead of grey -And thats the reason for the clicks/pops -right

**Comment by LeoBoudet** [comment]
> That sounds like the typical clicking you get with the envelope disabled. Did you click to enable the envelope before setting it? The envelope should be green instead of grey.

Yes I did. Some settings reduce the clicking noise like lowering the attack and the release but I have to completely deform the sound to make it disappear. Impossible to have an impact without having the clicking, even a small one.
On the other hand I notice that the crackling noise come and go for no apparent reason, sometimes discrete, sometimes awful. I suspect two different problems on my computer/LMMS, something relative to Linux in the case of the crackles. If I manage to get rid of it, I'll give the solution here.

**Comment by SecondFlight** [comment]
If you can reproduce the crackling you're describing in a demo project, let me know.

As for envelopes, it is possible to set them so they're not clicking but still have little impact. Could you send an example of how you have it set to reduce clicking?

**Comment by LeoBoudet** [comment]
Ok, I tried on debian buster and the problems are similar. But I manage to get a quite good sound from LMMS. I just hear some remaining crackling sounds while playing live many notes with the software but it is highly unlikely that I use so many notes at the same time in my projects.
Here are two screenshots of the amount of attack I have to set on two instruments:
![screenshots-lmms](https://user-images.githubusercontent.com/32602517/76685344-2e92b600-6613-11ea-92ca-c8e1ebdd7856.jpeg)
Basically, I have to low the attack a bit to get rid of the clipping (perfectly normal).

About the awful crackling noises I look further with some tests and it appears that ALSA is the responsible of this, not LMMS at all (selecting SDL solves the problem, at least in LMMS).

If you think that my settings in the screenshots are ok, you can close the issue, or I will, if necessary.

Thanks for your support.

Léo

**Comment by DouglasDGI** [comment]
> I just hear some remaining crackling sounds while playing live many notes with the software but it is highly unlikely that I use so many notes at the same time in my projects.

@LeoBoudet Does it still do that if you turn LMMS's volume down?  In all music-making programs (no exceptions), any audio that exceeds a volume above 0 dbFs will have their waveforms clipped down to fit within the bounds required, which causes distortion.  This is called "clipping".  You can tell your audio is clipping if the waveform visualizer at the top of the window turns red at any point.

**Comment by musikBear** [comment]
For almost 15 y ago the decision for the default envelope was made. Instead of a musical-sensible envelope, it was an envelope that displays all the options of an envelope, that was favoured. In the context of 'show-all' that does make sense, but it does definitively not in context of music!
This is an envelope of a real music-instrument:
![image](https://user-images.githubusercontent.com/6662066/76687463-7110be80-6624-11ea-8840-3a8ddf59bdb6.png)
You will find this in a guitar or piano.
You should study the chapter on envelopes : https://lmms.io/wiki/index.php?title=Instrument_Editor#Envelopes -Then you will learn how to use them correctly, and how to make different kinds of envelopes.
Adding a filter, also helps the output. Most producers will add *All-pass* filters to all instruments, even if they do not intend to automate the filtered  sound.
First read the chapter, then do the build-exercises, and do watch out for clipping!

**Comment by LeoBoudet** [comment]
I have no waveform visualizer, just the master volume oscillator, and even if it's full green I hear some crackling artefacts while playing the notes. Not the most inconvenient nevertheless.

I succeeded to obtain some good attacks without clicks in the entrance nor in the exit of the notes.

Now my main problem comes from ALSA. I'll investigate further in the next days and leave note here, in case it's relevant.

Many thanks for your time.

**Comment by LeoBoudet** [comment]
Just to illustrate what I am talking about, I attached an audio where I play with volume variation of each note (no saturation indicated, everything is green). The noise only shows up when I move alongside the piano keys.
I suppose it comes from the audio management of Linux and not from LMMS but I let it here, just in case.

[crackling-volume-variations.mp3.zip](https://github.com/LMMS/lmms/files/4334235/crackling-volume-variations.mp3.zip)

**Comment by musikBear** [comment]
@LeoBoudet I can hear it -yes, 
I wonder if you should drop ALSA and use SDL. 
Does you get same artefact with SDL
Create a project with notes just like those you have here, and confirm that you have ALSA-crackling, then change back-end to SDL and play the *same* project -Witch backend has the cleanest output?
Please upload that project here too!

About the artefact itself..
I think it is caused by resonance in excess. How close are you at clip-value?
You could insert an EQ, and turn down the lowest registers, i think that also would remove the crackles

**Comment by LeoBoudet** [comment]
> @LeoBoudet I can hear it -yes,
I wonder if you should drop ALSA and use SDL.
Does you get same artefact with SDL
Create a project with notes just like those you have here, and confirm that you have ALSA-crackling, then change back-end to SDL and play the same project -Witch backend has the cleanest output?
Please upload that project here too!

That's interesting, for this problem in particular, I only have it when I play live with the piano, but not if I create a project. Same results with both SDL and ALSA. If I want to completely get rid of the crackling noise while playing with the piano keys, I have to completely destroy the sound, like this:
![triple-oscillator-eq](https://user-images.githubusercontent.com/32602517/76704657-d2936480-66da-11ea-8175-8f2edb71f5a1.png)
However, I tried several sound tests on a project through automation or simply by playing A LOT of notes simultaneously, and the sound was ok.

The awful crackling that I experienced before didn't show up again in Ubuntu 18.04.4 LTS (it came up in debian buster yesterday). This one seems erratic and is not caused by LMMS (I had it with different softwares).

I will go further into LMMS as it seems to be a really powerful software. I don't know if other people noticed some noise while playing live with the piano keys so for now, I suggest to let the issue opened after all. What do you think?

**Comment by SecondFlight** [comment]
> I just hear some remaining crackling sounds while playing live many notes with the software

This is likely expected. If you enable oscilloscope, you will see red if the sound is [clipping](https://en.wikipedia.org/wiki/Clipping_(audio)).

**Comment by SecondFlight** [comment]
Could you check if the crackling you hear when playing live is helped by turning down the volume of the instrument you're playing?

**Comment by LeoBoudet** [comment]
It does exactly the same but at a lower volume. This doesn't seem to be a clipping effect: everything is green, no saturation observed.

**Comment by musikBear** [comment]
@LeoBoudet As i said
> Please upload that project here too!

I cant do a lot without something to do something **with**  -Right?
So upload the project or a demo with the fault

**Comment by LeoBoudet** [comment]
@musikBear As I said

> That's interesting, for this problem in particular, I only have it when I play live with the piano, but not if I create a project. Same results with both SDL and ALSA.

See last attached audio to hear what I experience.

**Comment by musikBear** [comment]
> See last attached audio to hear what I experience.

I need a demo project

**Comment by PhysSong** [comment]
@LeoBoudet If there are any problems with audio playback, could you try increasing the buffer size in the settings dialog?
